src/libawkward/io/json.cpp

```diff
@@ -540,44 +622,190 @@ namespace awkward {
+  {
+    bool scan = true;
+    bool has_error = false;
+    while (scan) {
+      scan = false;
+      reader.Parse<rj::kParseStopWhenDoneFlag>(stream, handler);
+      if(stream.Peek() == '\n') {
+        scan = true;
+      } else if (stream.Peek() == '\r') {
+        scan = true;
+      } else if (stream.Peek() != 0) {
+        has_error = true;
+      }
+    }
+    if (has_error) {
       throw std::invalid_argument(
-        std::string("JSON error at char ")
-        + std::to_string(reader.GetErrorOffset()) + std::string(": ")
-        + std::string(rj::GetParseError_En(reader.GetParseErrorCode()))
+        std::string("JSON File error at char ")
+        + std::to_string(stream.Tell()) + std::string(": \'")
+        + stream.Peek() + std::string("\'")
         + FILENAME(__LINE__));
     }
+    return handler.snapshot();
+  }
+
+  const ContentPtr
+  FromJsonString(const char* source,
+                 const ArrayBuilderOptions& options,
+                 const char* nan_string,
+                 const char* infinity_string,
+                 const char* minus_infinity_string,
+                 bool nan_and_inf_as_float) {
+    rj::Reader reader;
+    rj::StringStream stream(source);
+    if(nan_and_inf_as_float  ||  nan_string != nullptr
+      ||  nan_string != nullptr  ||  nan_string != nullptr) {
+      HandlerNanAndInf handler (options, nan_string, infinity_string,
+        minus_infinity_string, nan_and_inf_as_float);
+      return do_parse(handler, reader, stream);
+    } else {
+      Handler handler (options);
+      return do_parse(handler, reader, stream);
+    }
+    auto where = reader.GetErrorOffset();
+    throw std::invalid_argument(
+      std::string("JSON error at char ")
+      + std::to_string(where) + std::string(": \"") + source[where] + std::string("\" ")
+      + std::string(rj::GetParseError_En(reader.GetParseErrorCode()))
+      + FILENAME(__LINE__));
   }
 
   const ContentPtr
   FromJsonFile(FILE* source,
                const ArrayBuilderOptions& options,
-               int64_t buffersize) {
-    Handler handler(options);
+               int64_t buffersize,
+               const char* nan_string,
+               const char* infinity_string,
+               const char* minus_infinity_string,
+               bool nan_and_inf_as_float) {
     rj::Reader reader;
     std::shared_ptr<char> buffer(new char[(size_t)buffersize],
                                  kernel::array_deleter<char>());
     rj::FileReadStream stream(source,
                               buffer.get(),
                               ((size_t)buffersize)*sizeof(char));
-    if (reader.Parse(stream, handler)) {
-      return handler.snapshot();
+    if (nan_and_inf_as_float  ||  nan_string != nullptr
+      ||  nan_string != nullptr  ||  nan_string != nullptr) {
+      HandlerNanAndInf handler(options, nan_string, infinity_string,
+        minus_infinity_string, nan_and_inf_as_float);
+      return do_parse(handler, reader, stream);
```

Comment:
Right, it's the logical-or of all of those. Someone could set `nan_string` but not `infinity_string`, etc.